(WIP) Make fetch happen

Manishearth opened the pull request:
This adds a FetchTaskTarget, but doesn't use it yet. Further commits will make
XHR use the fetch task target.

@jdm thoughts so far?

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11439)

<!-- Reviewable:end -->

Manishearth committed:
Revamp Fetch async handling to use a fetch target and done channels

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net_traits/response.rs, components/net_traits/response.rs, components/net/fetch/methods.rs, components/net_traits/request.rs, components/net_traits/request.rs, components/net_traits/lib.rs, components/net_traits/lib.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify net code, but no tests are modified. Please consider adding a test!

KiChjang commented:
I suggest trying to get the unit tests working. That should give us a good idea about where your design is headed.

KiChjang commented on components/net_traits/request.rs:
```diff
@@ -106,6 +107,25 @@ pub enum CORSSettings {
     UseCredentials
 }
 
+pub trait FetchTaskTarget {
```
I would arguably put this in `net_traits/lib.rs`, `use net_traits::request::FetchTaskTarget` doesn't quite appeal to me.

highfive commented:
New code was committed to pull request.

Manishearth commented:
Tests work now (there was a minor issue where it hangs on network errors, fixed)

highfive commented:
New code was committed to pull request.

jdm commented:
Doesn't build :)

highfive commented:
New code was committed to pull request.

jdm commented:
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 4 of 4 files at r1.
Review status: 1 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 260 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImB0AtOxy-8ezWSs9N:-KImB0AuQ1YAL1xMZ-q-:-246401822) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L260)):*
> ```Rust
>     // Step 18
>     if request.synchronous {
>         done_chan.as_ref().map(|ref ch| ch.1.recv());
> ```

if let, although we probably want to assert that we actually have a done_chan here?

---

*[components/net/fetch/methods.rs, line 266 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImBDnY4uqvo-hDFUvV:-KImBDnY4uqvo-hDFUvW:-1539972643) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L266)):*
> ```Rust
>     // Step 19
>     if request.body.borrow().is_some() && matches!(request.current_url().scheme(), "http" | "https") {
>         target.as_mut().map(|target| {
> ```

if let

---

*[components/net/fetch/methods.rs, line 278 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImBFKUBbXaBy2HY-WI:-KImBFKUBbXaBy2HY-WJ:-1539972643) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L278)):*
> ```Rust
>     {
>         // Step 20
>         target.as_mut().map(|target| target.process_response(&response));
> ```

if let

---

*[components/net/fetch/methods.rs, line 281 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImBW4Ye7CXH8axF5va:-KImBW4Ye7CXH8axF5vb:-1581754749) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L281)):*
> ```Rust
> 
>         // Step 21
>         done_chan.as_ref().map(|ref ch| ch.1.recv());
> ```

We specifically need to wait for the internal response to be complete, and I find it difficult to reason about whether that's actually happening now.

---

*[components/net/fetch/methods.rs, line 284 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImBiv9RDdMLYNM8ks4:-KImBivAqn0BR1BE4iIg:-1539972643) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L284)):*
> ```Rust
> 
>         // Step 22
>         target.as_mut().map(|t| t.process_response_eof(&response));
> ```

if let

---

*[components/net/fetch/methods.rs, line 445 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImBrpZx51HZTwLaby3:-KImBrpZx51HZTwLaby4:-1548412952) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L445)):*
> ```Rust
>         // Substep 4
>         let fetch_result = http_network_or_cache_fetch(request.clone(), credentials, authentication_fetch_flag,
>                                                        done_chan.as_ref().map(|ref chan| chan.0.clone()));
> ```

I'd rather have a separate local variable than have this inline here.

---

*[components/net/fetch/methods.rs, line 472 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImBxpj7soeK092UzcE:-KImBxpkTqbePDe5eH3Q:1950323870) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L472)):*
> ```Rust
>                 },
>                 RedirectMode::Follow => {
> ```

Let's remove this newline.

---

*[components/net/fetch/methods.rs, line 883 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImC8qDze1qI8DKxrvV:-KImC8qDze1qI8DKxrvW:-1539972643) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L883)):*
> ```Rust
>                             };
>                             *res_body.lock().unwrap() = ResponseBody::Done(completed_body);
>                             done_sender.map(|s| s.send(()));
> ```

if let

---

*[components/net/fetch/methods.rs, line 893 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImCBIGmzrVsj4X6y_D:-KImCBIGmzrVsj4X6y_E:-1539972643) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L893)):*
> ```Rust
>         Err(_) => {
>             response.termination_reason = Some(TerminationReason::Fatal);
>             done_sender.map(|s| s.send(()));
> ```

if let

---

*[components/net_traits/request.rs, line 115 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImCVntrbIx94JDn0lh:-KImCVntrbIx94JDn0li:-1465209885) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net_traits/request.rs#L115)):*
> ```Rust
>     /// Fired when a chunk of the request body is transmitted
>     fn process_request_body(&mut self, request: &Request);
>     /// https://fetch.spec.whatwg.org/#process-request-end-of-file
> ```

Add a newline after each method, please.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11439#-:-KImCyfypyBzGWeruqRT:1458266033)*
<!-- Sent from Reviewable.io -->

highfive commented:
New code was committed to pull request.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #11472) made this pull request unmergeable. Please resolve the merge conflicts.

Manishearth committed:
Net side of XHR fetch integration

jdm commented:
I still haven't thought hard about what waiting on the channel means, so leaving the awaiting-review label.
+S-needs-code-changes +S-fails-tidy 

<details><summary><i>Previously, bors-servo wrote…</i></summary>
> :umbrella: The latest upstream changes (presumably #11472) made this pull request unmergeable. Please resolve the merge conflicts.
</details>


---

Reviewed 4 of 4 files at r2.
Review status: 2 of 5 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[components/net/fetch/methods.rs, line 261 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11439#-KJ1hlxjIsZxVMxMkzL6:-KJ1hlxjIsZxVMxMkzL7:1024930632) ([raw file](https://github.com/servo/servo/blob/fe71d4afa35a77c39327ca0ab21f5cf889e742b1/components/net/fetch/methods.rs#L261)):*
> ```Rust
>     // Step 18
>     if request.synchronous {
>         if !response.is_network_error() {
> ```

Is there a test for this?

---

*[components/net_traits/lib.rs, line 160 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11439#-KJ1iBBDZvEIGU8M5uDr:-KJ1iBBDZvEIGU8M5uDs:-1465209885) ([raw file](https://github.com/servo/servo/blob/fe71d4afa35a77c39327ca0ab21f5cf889e742b1/components/net_traits/lib.rs#L160)):*
> ```Rust
>     ///
>     /// Fired when a chunk of the request body is transmitted
>     fn process_request_body(&mut self, request: &request::Request);
> ```

Add a newline after each method, please.

---

*[tests/unit/net/fetch.rs, line 49 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11439#-KJ1iTq1Oja65e6pvgUP:-KJ1iTq25glU5p6la0rL:1653883965) ([raw file](https://github.com/servo/servo/blob/fe71d4afa35a77c39327ca0ab21f5cf889e742b1/tests/unit/net/fetch.rs#L49)):*
> ```Rust
> }
> 
> fn fetch_async(request: Request, target: Box<FetchTaskTarget+Send>) {
> ```

This looks like fetch_async from methods.rs. We should duplicate the useful pieces, like the thread name.

---

*[tests/unit/net/fetch.rs, line 80 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11439#-KJ1jEP1wbchYukqQ2fI:-KJ1jEP1wbchYukqQ2fJ:-1717091122) ([raw file](https://github.com/servo/servo/blob/fe71d4afa35a77c39327ca0ab21f5cf889e742b1/tests/unit/net/fetch.rs#L80)):*
> ```Rust
>     let wrapped_request = Rc::new(request);
> 
>     let fetch_response = fetch(wrapped_request, None);
> ```

We should define a `fetch_sync` method that hides the None argument, and not leave fetch publicly exposed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11439#-:-KJ1jZ3Js7z2luvMTnMQ:1211718575)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 2 of 5 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[components/net/fetch/methods.rs, line 260 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11439#-KImB0AtOxy-8ezWSs9N:-KJ5BMsbIc8MdCuO0hDN:-485225153) ([raw file](https://github.com/servo/servo/blob/1e3e246c05c3bf5448d72aae16f65476c6740ab4/components/net/fetch/methods.rs#L260)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> if let, although we probably want to assert that we actually have a done_chan here?
</details>
No, done_chan can be None in any of the cases where the request doesn't need a thread to complete (cache, network error, etc)

---

*[components/net/fetch/methods.rs, line 261 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11439#-KJ1hlxjIsZxVMxMkzL6:-KJ5BJqn9TKhYUXXG-A0:753841247) ([raw file](https://github.com/servo/servo/blob/fe71d4afa35a77c39327ca0ab21f5cf889e742b1/components/net/fetch/methods.rs#L261)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Is there a test for this?
</details>
For what?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11439)*
<!-- Sent from Reviewable.io -->

jdm commented:
Reviewed 1 of 1 files at r3.
Review status: 2 of 5 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[components/net/fetch/methods.rs, line 261 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11439#-KJ1hlxjIsZxVMxMkzL6:-KJ5m2_2jT8l9RBZwCqW:966776736) ([raw file](https://github.com/servo/servo/blob/fe71d4afa35a77c39327ca0ab21f5cf889e742b1/components/net/fetch/methods.rs#L261)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> For what?
</details>
For this new code that was added to handle an error case. What was the motivation for adding it, and can we write a test for it?

---

*[components/net/fetch/methods.rs, line 262 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11439#-KJ5mFU0Sm6iZ50DQHT5:-KJ5mFU0Sm6iZ50DQHT6:-1539972643) ([raw file](https://github.com/servo/servo/blob/fe71d4afa35a77c39327ca0ab21f5cf889e742b1/components/net/fetch/methods.rs#L262)):*
> ```Rust
>     if request.synchronous {
>         if !response.is_network_error() {
>             done_chan.as_ref().map(|ref ch| ch.1.recv());
> ```

if let

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11439)*
<!-- Sent from Reviewable.io -->

Manishearth committed:
Abstractify the Listener abstractions further

Manishearth committed:
Hook XHR up to fetch

Manishearth committed:
Use RequestInit instead of LoadData. Make code look like the spec.

Manishearth commented:
Most tests seem to pass, but I'm getting timeouts (which I'll go through tomorrow)

```
$ ./mach test-wpt --include tests/wpt/web-platform-tests/XMLHttpRequest/
Running 198 tests in web-platform-tests

Unsupported test type wdspec for product servo
  ▶ TIMEOUT [expected OK] /XMLHttpRequest/XMLHttpRequest-withCredentials.worker

STDERR: 127.0.0.1 - - [01/Jun/2016 18:48:49] code 400, message Bad request syntax ('0')
STDERR: 127.0.0.1 - - [01/Jun/2016 18:48:49] "0" 400 -
  ▶ TIMEOUT [expected OK] /XMLHttpRequest/event-readystatechange-loaded.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/interfaces.html

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/open-url-about-blank-window.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/open-url-multi-window-2.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/open-url-multi-window-3.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/open-url-multi-window-4.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/open-url-multi-window-5.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/open-url-multi-window.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/response-data-deflate.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/response-data-progress.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/response-invalid-responsetype.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/responsetype.html

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/responseurl.html

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/responsexml-document-properties.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/responsexml-media-type.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-authentication-basic-repeat-no-args.htm
  │ 
  └ Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-authentication-basic-setrequestheader-existing-session.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-authentication-basic.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-content-type-string.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-entity-body-document.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-send.worker

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-timeout-events.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-usp.html

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-usp.worker

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/status-async.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/status-basic.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/status-error.htm

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/xmlhttprequest-network-error.htm

Ran 198 tests finished in 405.0 seconds.
  • 169 ran as expected. 2 tests skipped.
  • 29 tests timed out unexpectedly
```

Manishearth committed:
Test fixes:

 - Hack to stop hitting unreachable on referer
 - add fetch_done to make sync work
 - Make datauris work by setting the response URI, spec bug
 - Allow for empty bodies
 - Make request bodies work (pass to http, fix fencepost in iter count)

Manishearth commented:
Sync seems to work now, though we still have behavior-related test failures

http://hastebin.com/lixazoyowo.txt

Ran 198 tests finished in 114.0 seconds.
  • 172 ran as expected. 2 tests skipped.
  • 2 tests timed out unexpectedly
  • 26 tests had unexpected subtest results

Manishearth commented:
Now at http://hastebin.com/lequyacuzi.txt, 3 tests fixed
